File: cpp/queue.cpp

#include <iostream>
#include <thread>
#include <bitset>

struct Node
{
    int val;
    Node* next;
    Node(int a) : val(a), next(nullptr) {}
};

struct Queue
{
    Node* front;
    Node* rear;

    Queue() : front(nullptr), rear(nullptr) {}

    bool empty()
    {
        return !front;
    }

    void Enqueue(int a)
    {
        Node* node = new Node(a);
        if (empty())
        {
            front = rear = node;
        }
        else
        {
            rear->next = node;
            rear = node;
        }
    }

    int Dequeue()
    {
        if (empty())
        {
            std::cout << "Queue is empty\n";
        }
        else
        {
            Node* tmp = front;
            front = tmp->next;
            int out = tmp->val;
            if (!front)
                rear = nullptr;
            delete tmp;
            return out;
        }
    }

    int PeekFront()
    {
        return front->val;
    }

    int PeekRear()
    {
        return rear->val;
    }

    void Print()
    {
        if (empty())
        {
            std::cout << "Queue is empty\n";
        }
        else
        {
            std::cout << front->val << " <- front\n";
            Node* tmp = front->next;
            while (tmp)
            {
                std::cout << tmp->val;
                std::cout << (tmp->next != nullptr ? "\n" : " <- rear\n");
                tmp = tmp->next;
            }
        }
    }

    ~Queue()
    {
        while(!empty())
        {
            Dequeue();
        }
    }
};


int main()
{
    Queue line;

    line.Enqueue(400);
    line.Enqueue(500);
    line.Enqueue(600);
    line.Enqueue(700);
    line.Print();
    
    std::cout << "\n";
    int out = line.Dequeue();
    std::cout << out << " <- dequeue'd\n\n";

    line.Print();
}

File: python/queue.py
# would this also work?
import sys

class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    def Enqueue(self, val):
        node = Node(val)
        if not self.front:
            self.front = self.rear = node
        else:
            self.rear.next = node
            self.rear = node

    def Dequeue(self):
        if not self.front:
            print("Queue is empty")
            return
        out = self.front.val
        self.front = self.front.next
        if self.front is None:
            self.rear = None
        return out
    
    def Print(self):
        if not self.front:
            print("Queue is empty")
            return
        print(f"{self.front.val} <- front")
        tmp = self.front.next
        while tmp:
            if not tmp.next:
                print(f"{tmp.val} <- rear")
                break
            print(tmp.val)
            tmp = tmp.next

q = Queue()

q.Enqueue(40)
q.Enqueue(50)
q.Enqueue(60)
q.Enqueue(70)

q.Print()
            